File: init/NM_10t/May2024/Specified_ne_COSMIC/perturb.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Aug 25 13:12:56 2024

@author: zettergm
"""

import typing as T
import numpy as np
from numpy import tanh
import scipy

from neprofile import readprofile
import gemini3d.read
import gemini3d.write

def perturb(cfg: T.Dict[str, T.Any], xg: T.Dict[str, T.Any]):
    # Data from HF systems
    filename="fp_profile.txt"
    z,ne = readprofile(filename)
    ne=ne*1.4
    
    # WARNING total hack; scale up the profile to compensate for decay during
    #   the "settling"/staging simulation
    #ne = 1.4 * ne
    
    # need to do something about out-of-bounds altitudes
    zmin=z.min()
    zmax=z.max()
    zgridmin=xg["alt"].min()
    zgridmax=xg["alt"].max()
    lsample=512
    zdatasample=np.linspace(zgridmin,zgridmax,lsample)
    i=0
    nedatasample=np.zeros(zdatasample.shape)
    while zdatasample[i]<zmin:
        nedatasample[i]=1e-20
        i+=1
    imin=i
    while i<lsample and zdatasample[i]<zmax:
        i+=1
    imax=i
    nedatasample[imin:imax]= scipy.interp(zdatasample[imin:imax],z,ne[:])
       
    for i in range(imax,lsample):
        ne3=nedatasample[i-3]
        ne2=nedatasample[i-2]
        nedatasample[i]=nedatasample[i-1]*ne2/ne3
               
    # construct a full initial condition for density of all species from total
    #   plasma density + assumptions
    x1 = xg["x1"][2:-2]
    x2 = xg["x2"][2:-2]
    x3 = xg["x3"][2:-2]
    nsperturb = np.empty((7, x1.size, x2.size, x3.size))
    for i in range(0, x2.size):
        for j in range(0, x3.size):
            nsperturb[6, :, i, j] = scipy.interp(xg["alt"][:,i,j],zdatasample,nedatasample[:])

    # apply some assumed composition (doesn't matter much unless chemistry is used)
    comp = 1 / 2 + 1 / 2 * tanh((x1 - 200e3) / 15e3)            
    for i in range(0, x2.size):
        for j in range(0, x3.size):            
            nmolec = (1 - comp) * nsperturb[6, :, i, j]
            natomic = comp * nsperturb[6, :, i, j]
            nsperturb[0, :, i, j] = 0.98 * natomic
            nsperturb[4, :, i, j] = 0.01 * natomic
            nsperturb[5, :, i, j] = 0.01 * natomic
            nsperturb[1, :, i, j] = 1 / 3 * nmolec
            nsperturb[2, :, i, j] = 1 / 3 * nmolec
            nsperturb[3, :, i, j] = 1 / 3 * nmolec

    # enforce nonzero min density
    nsperturb = np.maximum(nsperturb, 1e4)
    nsperturb[6, :, :, :] = np.sum(nsperturb[0:5, :, :, :], axis=0)
    
    ###########################################################################
    # Write in a format GEMINI can handle
    ###########################################################################
    # read in original reference data; we will retain flows and temperatures
    dat = gemini3d.read.frame(cfg["indat_file"], var=["ns", "Ts", "vs1"])
    
    gemini3d.write.state(
    cfg["indat_file"],
    dat,
    ns=nsperturb
    )    

File: init/NM_10t/May2024/Specified_ne_COSMIC/perturb.m
function perturb(cfg, xg)
    % Constants
    filename = "fp_profile.txt";

    % Read the plasma profile
    [z, ne] = neprofile(filename);
    %ne = ne * 1.4;

    % Out-of-bounds altitude handling
    zmin = min(z);
    zmax = max(z);
    zgridmin = min(xg.alt(:));
    zgridmax = max(xg.alt(:));
    lsample=512;
    zdatasample = linspace(zgridmin, zgridmax, lsample);
    
    % Initialize sampled density
    nedatasample = zeros(size(zdatasample));
    i = 1;
    
    % Fill data for points below min altitude of data
    while zdatasample(i) < zmin
        nedatasample(i) = 1e-20;
        i = i + 1;
    end
    imin = i;

    % Interpolation for points within data bounds
    while i <= lsample && zdatasample(i) < zmax
        i = i + 1;
    end
    imax = i-1;
    nedatasample(imin:imax) = interp1(z, ne, zdatasample(imin:imax), 'linear', 'extrap');
    
    % Fill value for data above max altitude range
    for i = imax+1:lsample
        ne3 = nedatasample(i-3);
        ne2 = nedatasample(i-2);
        nedatasample(i) = nedatasample(i-1) * ne2 / ne3;
    end

    % Construct initial condition for density of all species
    x1 = xg.x1(3:end-2);
    x2 = xg.x2(3:end-2);
    x3 = xg.x3(3:end-2);
    nsperturb = zeros(7, length(x1), length(x2), length(x3));

    for i = 1:length(x2)
        for j = 1:length(x3)
            nsperturb(7, :, i, j) = interp1(zdatasample, nedatasample, squeeze(xg.alt(:, i, j)), 'linear', 'extrap');
        end
    end

    % Apply assumed composition
    comp = 0.5 + 0.5 * tanh((x1 - 200e3) / 15e3);
    for i = 1:length(x2)
        for j = 1:length(x3)
            nmolec = (1 - comp') .* nsperturb(7, :, i, j);
            natomic = comp' .* nsperturb(7, :, i, j);
            nsperturb(1, :, i, j) = 0.98 * natomic;
            nsperturb(5, :, i, j) = 0.01 * natomic;
            nsperturb(6, :, i, j) = 0.01 * natomic;
            nsperturb(2, :, i, j) = (1/3) * nmolec;
            nsperturb(3, :, i, j) = (1/3) * nmolec;
            nsperturb(4, :, i, j) = (1/3) * nmolec;
        end
    end

    % Enforce nonzero minimum density
    nsperturb = max(nsperturb, 1e4);
    nsperturb(7, :, :, :) = sum(nsperturb(1:5, :, :, :), 1);

    % Read frame data
    % dat = gemini3d.read.frame(cfg.indat_file, var=["ns", "Ts", "vs1"]);
    dat = gemini3d.read.frame(cfg.indat_file,"vars",["ns", "Ts", "v1"]);

    % Write the new state
    % gemini3d.write.state(cfg.indat_file,dat,"ns",nsperturb);
    gemini3d.write.state(cfg.indat_file, dat);
end
